### PR TITLE
[PlayList] Fix resume on playlist STRM files

### LINKED ISSUE #19981: kodi fails to resume .strm files
### Describe the bug
you cannot resume  a movie in that is in a  .strm file

The movie in a .strm file can be stopped and resumed as long as you don't navigate away from the directory. Once you leave and re-enter the directory it is no longer possible to resume the movie

## Expected Behavior

when stopping the movie with the square button, and navigating away or closing/restarting kodi, it should be possible to resume the movie

## Actual Behavior

currently stopping the movie with the square button, and closing/restarting kodi, and trying to resume the movie, makes the movie start from the beginning

### To Reproduce

1. create a .strm file containing an url, e.g

``` https://archive.org/download/thekillers1946usafeaturingburtlancasteravagardneredmondobrienfilmnoirfullmovie_202001/The%20Killers%20%281946%2C%20USA%29%20Featuring%20Burt%20Lancaster%2C%20Ava%20Gardner%2C%20Edmond%20O%27Brien%20-%20Film%20Noir%20Full%20Movie.mp4
``` 
2. place this mymovie.strm in a source (local, smb, it does not matter...)
3. Navigate to the source Videos => Files => your source (notice the file shows as mymovie.strm)
4. launch the movie, fast forward to the middle
5. stop the movie with the square button
6. notice at this point that the url shows in the directory (rather than mymovie.strm) and if you try to resume the movie it works
7. leave the directory (go to the parent), then reenter the directory
8. notice that the movie is now called mymovie.strm and it has the icon showing that the movie is in progress
9. try to  resume the movie, the position is lost and it starts from the beginning

### Debuglog

The debuglog can be found here:
[debuglog.zip](https://github.com/xbmc/xbmc/files/6834719/debuglog.zip)

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [x] Android
 - [ ] iOS
 - [ ] tvOS
 - [x] Linux
 - [ ] OSX
 - [x] Windows
 - [ ] Windows UWP

 - Operating system version/name: Linux arch1 5.12.15-arch1-1 #1 SMP PREEMPT Wed, 07 Jul 2021 23:35:29 +0000 x86_64 GNU/Linux
 - Kodi version: 19.1 (19.1.0) Git:20210509-nogitfound Media Center Kodi

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by CastagnaIT**
## Description

This fix the bookmark resume playback on playlist STRM files
this PR superseed PR #16076, when time ago i tested old PR was not working good i dont know why or my fails or somelse changes over time, now i tried rebasing and this time has worked good, so thanks to @Acidzero2020 for his search

my tests i have done as follow:

#### Test 1: More STRM files in a folder like a tvshow episodes, no scraper
- add more strm in a folder to be add to kodi library without set scraper
- open first STRM, seek to middle then stop playback
- open second STRM, seek to middle then stop (before the fix, the first strm loose the resume)
- open again the first STRM, seek to middle then stop (before the fix, the second strm loose the resume)

#### Test 2: More STRM files in a folder like a tvshow episodes, with scraper
same steps as Test 1 but by set the tvdb scraper to the folder

#### Test 3: Test resume on a Widget "Recently added episodes"
- add new STRM files on the tvhsow folder added in kodi library and update library to make scraper works
- If Kodi not update the tv show window close Kodi and reopen it
- The category widget "Recently added episodes" should appears
- Play a STRM on the widget (see screenshot) seek to middle and stop playback
- Play same STRM or another one, before the fix the video resume will be lost and you cannot resume video

#### Test 4: Test persistent resume over Kodi app close
Try play some STRM and then check if the video resume is persistent after you close/reopen kodi

#### Test 5: STRM file with multiple links contained on same file
handled as folder, video resume not available, so behaviour unchanged

## Motivation and context

Bookmark resume on playlist files like STRM never worked on kodi
the use of STRM with kodi library is widespread in addons
and all addons have had to implement various kind of workarounds
like reimplementing video resume bookmark by sniffing kodi library database or by using an own internal addon database where save the bookmark manually

fix #19981

## Screenshots (if appropriate):
immagine

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [x] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/playlists/PlayList.cpp [hunk 1]
```diff
@@ -476,6 +476,7 @@ bool CPlayList::Expand(int position)
   {
     (*playlist)[i]->SetDynPath((*playlist)[i]->GetPath());
     (*playlist)[i]->SetPath(item->GetPath());
+    (*playlist)[i]->m_lStartOffset = item->m_lStartOffset;
```
Comment (enen92) [inline xbmc/playlists/PlayList.cpp:479]: I would take the opportunity to create a setters and getters for the offset

File: xbmc/video/windows/GUIWindowVideoBase.cpp [hunk 2]
```diff
@@ -581,7 +581,7 @@ void CGUIWindowVideoBase::GetResumeItemOffset(const CFileItem *item, int64_t& st
   startoffset = 0;
   partNumber = 0;
 
-  if (!item->IsNFO() && !item->IsPlayList())
+  if ((!item->IsNFO() && !item->IsPlayList()) || item->IsType(".strm"))
```
Comment (enen92) [inline xbmc/video/windows/GUIWindowVideoBase.cpp:584]: This probably could be hidden in some new CFIleItem method (`IsResumable()`)

**Review by enen92 — APPROVED** [review]
LGTM. Your strm fix commit should come after the "cleanup" one and not the other way around but personally I don't see it as a hard requirement.

**Comment by CastagnaIT** [comment]
i add WIP because want try investigate for another use case that seem not works

**Comment by CastagnaIT** [comment]
found the problem, i was playling a short video and seeking within 3 minuts kodi skip saving the resume point as:
https://github.com/xbmc/xbmc/blob/master/xbmc/settings/AdvancedSettings.cpp#L159
so all ok

File: xbmc/FileItem.h [hunk 3]
```diff
@@ -362,6 +362,32 @@ class CFileItem :
...
+   *         Can return also special value -1, see define STARTOFFSET_RESUME.
+   */
+  int64_t GetStartOffset() const { return m_lStartOffset; }
+
+  /*!
+   * \brief Set the offset where start the playback.
+   * \param offset Set the offset value as ms,
+                   or the special value STARTOFFSET_RESUME.
+   */
+  void SetStartOffset(const int64_t offset) { m_lStartOffset = offset; }
```
Comment (thexai) [inline xbmc/FileItem.h:383]: `const int64_t& offset`

File: xbmc/FileItem.h [hunk 4]
```diff
@@ -362,6 +362,32 @@ class CFileItem :
...
+   * \brief Get the end offset.
+   * \return The offset value as ms.
+   */
+  int64_t GetEndOffset() const { return m_lEndOffset; }
+
+  /*!
+   * \brief Set the end offset.
+   * \param offset Set the offset as ms.
+   */
+  void SetEndOffset(const int64_t offset) { m_lEndOffset = offset; }
```
Comment (thexai) [inline xbmc/FileItem.h:395]: const int64_t& offset

File: xbmc/FileItem.h (hunk 3 above)
Comment (garbear) [reply xbmc/FileItem.h:383]: Rather `int64_t offset`, `const` shouldn't be used here because it exposes internal details (constness) to the outside world.

**Pull request merged**